test/apps/network/tcp_poll.c

```diff
@@ -0,0 +1,259 @@
+
+	pfd.fd = sk_listen;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT | POLLHUP));
+
+	sk_addr.sin_family = AF_INET;
+	sk_addr.sin_port = S_PORT;
+	CHECK(inet_aton("127.0.0.1", &sk_addr.sin_addr));
+	CHECK(bind(sk_listen, (struct sockaddr *)&sk_addr, sizeof(sk_addr)));
+
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT | POLLHUP));
+}
+END_TEST()
+
+FN_TEST(poll_listen)
+{
+	CHECK(listen(sk_listen, 3));
+
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == 0);
+}
+END_TEST()
+
+FN_TEST(poll_connect_close)
+{
+	sk_connect = CHECK(socket(PF_INET, SOCK_STREAM, 0));
+	CHECK(connect(sk_connect, (struct sockaddr *)&sk_addr,
+		      sizeof(sk_addr)));
+
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT));
+	pfd.fd = sk_listen;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN));
+
+	close(sk_connect);
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN));
+
+	struct sockaddr addr;
+	socklen_t addrlen = sizeof(addr);
+	int sk = CHECK(accept(sk_listen, &addr, &addrlen));
+	pfd.fd = sk;
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP));
+}
+END_TEST()
+
+FN_TEST(poll_connect_accept)
+{
+	sk_connect = CHECK(socket(PF_INET, SOCK_STREAM, 0));
+	CHECK(connect(sk_connect, (struct sockaddr *)&sk_addr,
+		      sizeof(sk_addr)));
+
+	pfd.fd = sk_listen;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN));
+
+	struct sockaddr addr;
+	socklen_t addrlen = sizeof(addr);
+	sk_accept = CHECK(accept(sk_listen, &addr, &addrlen));
+
+	pfd.fd = sk_accept;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT));
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT));
+	pfd.fd = sk_listen;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == 0);
+}
+END_TEST()
+
+FN_TEST(poll_read_write)
+{
+	CHECK(write(sk_accept, buf, 1));
+
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN | POLLOUT));
+
+	TEST_RES(read(sk_connect, buf, 4096), _ret == 1);
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT));
+
+	CHECK(write(sk_connect, buf, 4096));
+	CHECK(write(sk_connect, buf, 4096));
+
+	// Ensure all data is transmitted.
+	sleep(1);
+
+	pfd.fd = sk_accept;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN | POLLOUT));
+	CHECK(read(sk_accept, buf, 4096));
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN | POLLOUT));
+	CHECK(read(sk_accept, buf, 4096));
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT));
+}
+END_TEST()
+
+FN_TEST(poll_shutdown_read)
+{
+	CHECK(write(sk_connect, buf, 1));
+
+	pfd.fd = sk_accept;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLIN | POLLOUT));
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT));
+
+	CHECK(shutdown(sk_accept, SHUT_RD));
+	pfd.fd = sk_accept;
+	TEST_RES(poll(&pfd, 1, 0),
+		 pfd.revents == (POLLIN | POLLOUT | POLLRDHUP));
+	pfd.fd = sk_connect;
+	TEST_RES(poll(&pfd, 1, 0), pfd.revents == (POLLOUT));
+
+	CHECK(write(sk_connect, buf, 1));
+	CHECK(read(sk_accept, buf, 4096));
```

Comment:
I guess if you remove the `write`, the `read` will cause the aforementioned inifinte loop. So, another test case is needed.